File: plugins/OutcomeTesting.Plugins/GradingRules.cs
using System;

namespace OutcomeTesting.Plugins
{
    /// <summary>
    /// Pure decision logic for the AQS Judgement and Grading section: whether the primary
    /// root cause is owed and whether a recorded one must be let go (item 3, 2026-09-19).
    ///
    /// The rules about which grades the rest of the form still leaves available moved to
    /// ChecklistGating on 2026-09-22, when they stopped being about the Suitability core
    /// checks - and about the grade alone. Only GradeAllowedWithInsufficientEvidence stayed,
    /// because it names the grade's own scale and nothing else.
    ///
    /// Deliberately free of Dataverse types, like ResponseRules and CaseHeaderRules, so the
    /// rule can be tested without a fake organisation service. Mirrored client-side in
    /// app/src/features/reviews/gradingRules.ts and read by OT Review Detail; when one
    /// changes, change all three.
    ///
    /// The root cause answers "why was this not a pass". A passing file has no root cause to
    /// name, and Q-GR-02 is seeded mandatory, so before this rule every Pass review was held
    /// at submission by a question with no true answer - the checker had to invent one, and
    /// that invention then went out in the trend MI Q-GR-02 exists to feed (AD-022).
    /// </summary>
    public static class GradingRules
    {
        /// <summary>
        /// Advice Quality Grade. Load-bearing by name in OutcomeRules, which maps its answer
        /// to the case outcome, and in ChecklistGuards, which refuses to retire it.
        /// </summary>
        public const string GradeQuestionCode = "Q-GR-01";

        /// <summary>
        /// Primary root cause. Load-bearing by name from this rule onwards, which is why it
        /// joins the ChecklistGuards list (AD-122): retire it and every non-pass review
        /// loses the column the trend MI reads.
        /// </summary>
        public const string RootCauseQuestionCode = "Q-GR-02";

        /// <summary>
        /// The grade's own scale (Pass / Pass with issues / Insufficient evidence / Potential
        /// harm) and the nine-cause single select, each used by exactly one question in the
        /// V8 checklist - the grade scale by Q-GR-01 and the single select by Q-GR-02, which
        /// is what AD-055 deliberately arranged.
        ///
        /// Held here because both front ends already have the response type in hand where
        /// they do not have the question code: the portal writes it into data-response-type
        /// on every answer row, and the app reads it off the form row. They are a cheap first
        /// filter, never the last word - the plug-in confirms the question code before it
        /// clears anything, because a response type is a convention and a code is a contract.
        /// </summary>
        public const int GradeResponseType = 120910010;

        /// <summary>See <see cref="GradeResponseType"/>.</summary>
        public const int RootCauseResponseType = 120910003;

        /// <summary>
        /// The root causes as several ticks (project owner, 2026-09-24), which Q-GR-02 answers
        /// on from its successor version. The single select above stays recognised: reviews
        /// submitted before the change hold their one cause against the earlier version.
        /// </summary>
        public const int RootCausesResponseType = ResponseRules.TypeMultiSelectRootCause;

        /// <summary>Whether a question on this response type is the primary root cause.</summary>
        public static bool IsRootCauseResponseType(int responseType)
        {
            return responseType == RootCauseResponseType || responseType == RootCausesResponseType;
        }

        /// <summary>
        /// Whether this review still owes a primary root cause.
        ///
        /// Owed for every grade but Pass, including a grade outside the scale: an
        /// unrecognised value is not a pass, and demanding the root cause is the safe
        /// direction, exactly as OutcomeRules.TryGradeFromAnswer refuses to read one as a
        /// pass.
        ///
        /// Not owed while the grade is unanswered, and that is not the same as excusing it.
        /// The grade is mandatory in its own right, so an ungraded review is refused at
        /// submission anyway; what this avoids is naming Q-GR-02 in that refusal, when
        /// whether it is owed at all is not yet knowable. The checker answers the grade, and
        /// the root cause is asked for - or is not - on the next attempt.
        /// </summary>
        public static bool RootCauseRequired(int? gradeAnswer)
        {
            return gradeAnswer.HasValue && gradeAnswer.Value != ResponseRules.ChoicePass;
        }

        /// <summary>
        /// Whether a recorded primary root cause must be cleared, because the grade now says
        /// the file passed.
        ///
        /// Not simply the negation of <see cref="RootCauseRequired"/>: an unanswered grade
        /// owes no root cause and destroys none either. A checker who picks the root cause
        /// before the grade - the two sit one above the other and nothing stops it - would
        /// otherwise have the answer wiped on the way past.
        ///
        /// Clearing rather than hiding, because a hidden value is still a stored value: it
        /// would export as a live root cause against a passing case and read to anyone
        /// querying the table as a contradiction nobody could account for.
        /// </summary>
        public static bool RootCauseCleared(int? gradeAnswer)
        {
            return gradeAnswer.HasValue && gradeAnswer.Value == ResponseRules.ChoicePass;
        }

        /// <summary>
        /// The two grades a review may still carry once anything on it has been answered
        /// Insufficient evidence (item 10, 2026-09-19; widened 2026-09-22).
        ///
        /// Evidence the checker found insufficient on a core check cannot be reconciled with
        /// a file that passed, with or without issues. The grade is not overridden - the
        /// checker still chooses between Insufficient evidence and Potential harm, which is a
        /// judgement only they can make - it is the two grades that contradict their own ticks
        /// that stop being available.
        /// </summary>
        public static bool GradeAllowedWithInsufficientEvidence(int gradeAnswer)
        {
            return gradeAnswer == ResponseRules.ChoiceInsufficient
                || gradeAnswer == ResponseRules.ChoicePotentialHarm;
        }

    }
}

File: app/src/features/reviews/gradingRules.ts
/**
 * The AQS Judgement and Grading section's conditional primary root cause (item 3,
 * 2026-09-19).
 *
 * The client-side mirror of `GradingRules` in the plug-in assembly, which is the authority:
 * `SubmitReviewPlugin` decides there whether the root cause is owed, and
 * `ResponseProgressPlugin` clears a recorded one there when the grade returns to Pass, so
 * neither this page nor the portal can hold a different rule. This exists so the document
 * does not draw a question the review no longer owes (AD-041).
 *
 * When one changes, change all three - here, `GradingRules.cs`, and `OT Review Detail`.
 */

import type { FormRow, SectionedAnswer } from './reviewSections';

/**
 * Advice Quality Grade and Primary root cause. Both are read by name in compiled C# and
 * both are protected from retirement by `ChecklistGuards` (AD-122).
 */
export const GRADE_QUESTION_CODE = 'Q-GR-01';
export const ROOT_CAUSE_QUESTION_CODE = 'Q-GR-02';

/**
 * The grade's own four-value scale and the nine-cause single select. Each belongs to exactly
 * one question in the V8 checklist, which is what AD-055 deliberately arranged, so they are
 * a sound way to find the two rows on a page that has the response type in hand and may not
 * have the code.
 */
export const GRADE_RESPONSE_TYPE = 120910010;
export const ROOT_CAUSE_RESPONSE_TYPE = 120910003;

/** Pass, as `ResponseRules.ChoicePass` has it. */
const PASS = 120910300;

/**
 * Whether the review still owes a primary root cause.
 *
 * Owed for every grade but Pass, including a grade outside the scale - an unrecognised value
 * is not a pass, and asking for the cause is the safe direction.
 *
 * Not owed while the grade is unanswered. The grade is mandatory in its own right, so an
 * ungraded review is refused at submission anyway; until it is given, whether the root cause
 * is owed is simply not knowable.
 */
export function rootCauseRequired(gradeAnswer: number | null | undefined): boolean {
  return gradeAnswer != null && gradeAnswer !== PASS;
}

/**
 * Whether a recorded root cause must be let go, because the grade now says the file passed.
 *
 * Deliberately not the negation of `rootCauseRequired`: an unanswered grade owes no root
 * cause and destroys none either, so a checker who picks the cause before the grade does not
 * have it wiped on the way past. The clearing itself happens server-side; this is here so
 * both copies of the rule read the same way.
 */
export function rootCauseCleared(gradeAnswer: number | null | undefined): boolean {
  return gradeAnswer != null && gradeAnswer === PASS;
}

/** The answer recorded against a row, where the row's answer type carries one. */
function choiceOn(row: FormRow<SectionedAnswer>): number | null {
  /*
   * Read structurally rather than through `TickedAnswer`. `formBlocks` is generic over the
   * answer shape and the grade is a ticked answer whatever else the caller holds, so a cast
   * here is what lets the rule apply without narrowing every caller to one answer type.
   */
  const response = row.response as { answerChoice?: number | null } | null;
  return response?.answerChoice ?? null;
}

/** Whether the row is a question, matched on its response type and confirmed by its code. */
function isQuestion(
  row: FormRow<SectionedAnswer>,
  responseType: number,
  code: string,
): boolean {
  if (row.responseTypeValue !== responseType) return false;
  // An unplaced answer has no code to confirm against (`reviewSections`), and the response
  // type alone is enough there: it belongs to this question and no other.
  return row.code == null || row.code.trim().toUpperCase() === code;
}

/**
 * The rows to draw, with the primary root cause taken out unless this review owes one.
 *
 * Keyed on `rootCauseRequired`, so an **ungraded** review does not draw it either. The
 * batch's wording is "visible and required only when the Advice quality grade is anything
 * other than Pass", and nothing is not anything: the question appears when the grade
 * arrives, and only if that grade is not a Pass.
 *
 * `checklistDocument.test.ts` compares what is drawn here against the reference Checker
 * Checklist, and its fixture answers the grade for exactly this reason - a blank form has no
 * grade, so a conditional row cannot be read off one.
 *
 * Applied to a section's own rows, because the grade and the cause sit in the same section
 * and a grade found anywhere else would not be this section's. A section holding no grade
 * row is returned untouched - every section but Judgement and Grading, and on a Tax review
 * that one is not drawn at all.
 */
export function withoutUnaskedRootCause<T extends SectionedAnswer>(
  rows: FormRow<T>[],
): FormRow<T>[] {
  const grade = rows.find((row) => isQuestion(row, GRADE_RESPONSE_TYPE, GRADE_QUESTION_CODE));
  if (!grade) return rows;

  if (rootCauseRequired(choiceOn(grade))) return rows;

  return rows.filter(
    (row) => !isQuestion(row, ROOT_CAUSE_RESPONSE_TYPE, ROOT_CAUSE_QUESTION_CODE),
  );
}
